src/controllers/handlers/utils.ts

```diff
@@ -28,6 +26,53 @@ export async function validate<T extends string>(
   return {
     identity,
     realmName,
-    sceneId
+    sceneId,
+    parcel,
+    hostname
+  }
+}
+
+export function ensureSlashAtTheEnd(url: string): string | undefined {
+  if (!url) return undefined
+  return url.endsWith('/') ? url : `${url}/`
+}
+
+export function validateFilters(filters: { admin?: string }): {
+  valid: boolean
+  error?: string
+  value: { admin?: string }
+} {
+  if (filters.admin !== undefined && typeof filters.admin !== 'string') {
+    return {
+      valid: false,
+      error: 'admin must be a string',
+      value: {}
+    }
+  }
+
+  return {
+    valid: true,
+    value: {
+      admin: filters.admin ? filters.admin.toLowerCase() : undefined
+    }
+  }
+}
+
+export async function fetchBlacklistedWallets(blackListJson: string): Promise<Set<string>> {
+  try {
+    const response = await fetch(blackListJson)
+    if (!response.ok) {
+      throw new Error(`Failed to fetch deny list, status: ${response.status}`)
+    }
+    const data = await response.json()
+    if (data.users && Array.isArray(data.users)) {
+      return new Set(data.users.map((user: { wallet: string }) => user.wallet.toLocaleLowerCase()))
+    } else {
+      console.warn('Deny list is missing "users" field or it is not an array.')
+      return new Set()
+    }
+  } catch (error) {
+    console.error(`Error fetching deny list: ${(error as Error).message}`)
+    return new Set()
```

Comment:
returning empty seems wrong, it will by-pass validations when we cannot fetch the blacklist for any reason